app/services/order_cycle_distributed_products.rb

```diff
@@ -15,27 +15,45 @@ def initialize(order_cycle, distributor)
   #
   # @return [ActiveRecord::Relation<Spree::Product>]
   def relation
-    variants = order_cycle.variants_distributed_by(distributor)
-    products = variants.map(&:product).uniq
-
-    valid_products = products.reject do |product|
-      product_has_only_obsolete_master_in_distribution?(product, variants)
-    end
-    product_ids = valid_products.map(&:id)
-
-    Spree::Product.where(id: product_ids)
+    Spree::Product
+      .unscoped
+      .select('DISTINCT(spree_products.*)')
+      .joins(variants_including_master: :exchanges)
+      .merge(variants_distributed_by_distributor)
+      .joins(left_join_products_with_obsolete_master)
+      .where('products_with_obsolete_master.id IS NULL')
+      .where('spree_products.deleted_at IS NULL')
+      .where('variants_including_masters_spree_products.deleted_at IS NULL')
   end
 
   private
 
   attr_reader :order_cycle, :distributor
 
-  # If a product without variants is added to an order cycle, and then some variants are added
-  # to that product, but not the order cycle, then the master variant should not available for
-  # customers to purchase.
-  def product_has_only_obsolete_master_in_distribution?(product, distributed_variants)
```

Comment:
my main feedback is: there are no products with obsolete master in our DB, all products always have a "standard variant". we should enforce and trust that. Would that help simplify this query?

Reply:
the meaning of variant here is `master: false`, I know it's a bit cumbersome because we use the terms interchangeably.

Reply:
I am not sure we are talking about the same thing :-)

can you clarify this statement?
"the meaning of variant here is master: false"

My comment was about "If a product without variants is added to an order cycle", there's no such thing in OFN now, all products have at least one "standard variant" (master: false)

Reply:
Ah! that's an interesting plot twist. Could you point to the place in the codebase where we enforce that? just to see if we really can get rid of this PR and just remove this class. That would like a dream :tada:.

Reply:
first, tested in au, fr, uk and katuma, **products with no standard variant**, empty everywhere:
`select * from spree_products where not exists (select 1 from spree_variants where is_master = false and product_id = spree_products.id);
`

the code is in the product_decorator, ensure_standard_variant [here](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/models/spree/product_decorator.rb#L39)

there's even a migration doing it back in 2014 [here](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/db/migrate/20141003060622_add_standard_variant_to_products.rb)


SO, I looked into **exchanges with master variants**:
`select * from exchange_variants ev, spree_variants v where ev.variant_id = v.id and v.is_master = true;`

This query is empty in katuma and FR, the most recent instances.

In AU and UK, there are some old entries. In AU the latest is from 2015-05-22 and in UK the latest is from 2016-07-08.

The Order Cycles in au and uk with these master variants have closed around those dates as well, 2015 and 2016 respectively.

SO, looks like we can ignore that and move on to a world where OCs cannot have master variants :-)

what do you think?

Reply:
> SO, looks like we can ignore that and move on to a world where OCs cannot have master variants :-)

That's awesome! but we'll to create a separate data migration to deal with those old OCs in UK and AU.

Reply:
this is empty in both AU and UK:
`select v_master.id from spree_variants v_master where v_master.id in (select distinct v.id from exchange_variants ev, spree_variants v where ev.variant_id = v.id and v.is_master = true) and not exists (select id from spree_variants v where v.product_id = v_master.product_id and v.is_master = false);`

that means that all master variants in exchanges in AU and UK have **at least** one standard variant available. We can switch the master variant with one of the standard variants of the product. Because we are not changing order, only closed OCs I think this is fine.

Reply:
Sounds good. We need a migration to remove those master variants from old order cycles and then we can remove this code.